JVM per context

Ankit1010 opened the pull request:
Running each context in a separate process for issue #87

velvia commented:
Overall, looks really promising.   Some thoughts:

- Think about using roles for Akka Cluster.  The HTTP server would be one role, and all the JobManagerActors would be another role.
- We should really look at the JobDaoActor, and think about does this API make sense and scale.

velvia commented:
@Ankit1010 wait.  Only the `StoreJar` message carries jar bytes, and that is only from WebApi to the JobDaoManager, which is completely local, right?   I think that should be fine, not over the network.  I'm more concerned about starting jobs, I believe only the jar path is passed right?

velvia commented:
@Ankit1010 how is this going?  When I find some time I might hack on this myself.

Ankit1010 commented:
@velvia 
I will try to finish the rest this week. So here is what is remaining and some questions about your points
1. What is the concern about the JobDAOActor? Yes the StoreJar message is local. This is the message flow from JobManagerActor to JobDaoActor:
   
   i. starting job - GetLastUploadTime and GetJarPath (via job cache)
   ii. finishing job - JobFinished or JobErroredOut sent to JobStatusActor, both of which send SaveJobInfo to JobDAOActor

2. Akka roles are used - there are two roles, "manager" for JobManagerActors and "supervisor". The config option is added programmatically in JobServer/JobManager before creating the actor system. The AkkaClusterSupervisor then checks for the "manager" role before sending an Identify message. Should we be using this somewhere else? 

3. I will try using spark_submit instead of the manager_start script. I think that should work. If not, i can factor the common parts of the manager_start and server_start script out.

velvia commented:
@Ankit1010  thanks, that sounds great.

Re your questons:

1. JobDAOActor:  My concern was just sending all the bytes of the jar over the wire, since you had mentioned something about the frame size, which I assumed had to do with network transmission.

It sounds though like there is no network transmission of jar bytes over Akka, since WebApi -> JobDao is local, and JobDaoActor -> JobManagerActor is just the GetJarPath.

2.  I think that's fine. 

3. Sounds good.  If spark-submit works, I'd be interested in using it for job server overall, since it could help avoid some classpath / classloader issues folks are seeing.
4. Tests -- I'll think about it more after I look through more of the code, probably towards the end of the week.   My thoughts are using the MultiJvmTest framework, though I'm not sure how that would work with the Supervisor you are writing.  Basically something to simulate, what happens if the process dies or runs out of memory and can't respond?

Thanks again.

Ankit1010 commented:
@velvia 
I have been testing this and it seems to work. I think everything except tests are done. Sorry for the delay -  I will be away for a week or so further so i would appreciate any help with modifying the existing tests and writing new tests.

velvia commented:
@Ankit1010 Many thanks.  Will try to have a look and try it out this weekend.

madhujahagirdar commented:
It is possible to run the actors per context in a docker container ? If so we would like to contribute patch for it.

velvia commented:
Madhu,

Do you mean run each separate JVM process in its own docker container?
Hmmm.   I'd be interested in looking at such a patch.... as long as it's an
easy option, since not everyone would do that.

The other problem is that the jar loading works through a local file path.
As long as that's OK and you're not running the containers in Mesos or
something.

-Evan


On Fri, May 15, 2015 at 9:21 AM, madhujahagirdar <notifications@github.com>
wrote:

> It is possible to run the actors per context in a docker container ? If so
> we would like to contribute patch for it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spark-jobserver/spark-jobserver/pull/120#issuecomment-102452007>
> .
>



-- 
If you are free, you need to free somebody else.
If you have some power, then your job is to empower somebody else.
   --- Toni Morrison

madhujahagirdar commented:
We are planning to create a patch to store the jars in hdfs instead of local file. That would allow us to run on Mesos also.

velvia commented:
Sure, that would be very welcome.  Thanks.

-Evan

On Fri, May 15, 2015 at 9:55 AM, madhujahagirdar <notifications@github.com>
wrote:

> We are planning to create a patch to store the jars in hdfs instead of
> local file. That would allow us to run on Mesos also.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spark-jobserver/spark-jobserver/pull/120#issuecomment-102458052>
> .
>



-- 
If you are free, you need to free somebody else.
If you have some power, then your job is to empower somebody else.
   --- Toni Morrison

PierreVincent commented:
That's great stuff @Ankit1010, I tried this out and this seems to be working well and solving issues with multiple jobs running simultaneously (see #147)

Couple of comments though, as it seems there are some gaps in that Pull Request as to what is required to wire the whole thing to run properly. I understand these might be related to the work being in progress, but in case somebody else wants to try this out... and also to make sure that they're not forgotten about! 

I can add these comments inline in the files as well if that's better for you?

**Missing required configuration "deploy.manager-start-cmd"**

There is no default value in the configuration for this entry. I worked out from your own repo that this should be set to a path to bin/manager_start.sh

**manager_start.sh is not part of the deployed scripts**

With this change, the Job Server needs this file to start contexts, so it would make sense to make it part of the deployment.

**setenv.sh is used for both manager_start.sh and server_start.sh but checks if server is not already started**

This causes the manager_start.sh to fail when creating context (POST to /contexts fails).
Commenting out the check for existing PID works around the problem. Obviously this needs a better solution.

**akka issues with Oversize pay load when creating job**

Once I was able to create contexts, creating jobs started throwing errors like:

```
akka.remote.OversizedPayloadException: Discarding oversized payload [...] max allowed size 128000 bytes
```

Adding the following line in the Job Server configuration got rid of this problem:

```
akka.remote.netty.tcp.maximum-frame-size = 256000
```

Now, my knowledge in Akka is limited so I'm not sure if that's a recommended thing to do or if that error might point towards something that is done wrong when creating the job.

**Unable to run tests for assembly build**

When building the JAR, I had to disable the tests in assembly because a couple of them have compilation errors (JobManagerActorSpec and JobStatusActorSpec). I understand that you didn't work out yet how to set up the tests for the multi-JVM side of things, but it might be good to fix the existing tests in the build - or to explicitly ignore them for the time being, so that we can still run the rest of the tests in assembly.

velvia commented:
Pierre, many thanks for trying this out.  I'll be trying this out soon,
hopefully.

On Sat, May 16, 2015 at 12:40 AM, Pierre Vincent <notifications@github.com>
wrote:

> That's great stuff @Ankit1010 <https://github.com/Ankit1010>, I tried
> this out and this seems to be working well and solving issues with multiple
> jobs running simultaneously (see #147
> <https://github.com/spark-jobserver/spark-jobserver/issues/147>)
>
> Couple of comments though, as it seems there are some gaps in that Pull
> Request as to what is required to wire the whole thing to run properly. I
> understand these might be related to the work being in progress, but in
> case somebody else wants to try this out... and also to make sure that
> they're not forgotten about!
>
> I can add these comments inline in the files as well if that's better for
> you?
>
> *Missing required configuration "deploy.manager-start-cmd"*
>
> There is no default value in the configuration for this entry. I worked
> out from your own repo that this should be set to a path to
> bin/manager_start.sh
>
> *manager_start.sh is not part of the deployed scripts*
>
> With this change, the Job Server needs this file to start contexts, so it
> would make sense to make it part of the deployment.
>
> *setenv.sh is used for both manager_start.sh and server_start.sh but
> checks if server is not already started*
>
> This causes the manager_start.sh to fail when creating context (POST to
> /contexts fails).
> Commenting out the check for existing PID works around the problem.
> Obviously this needs a better solution.
>
> *akka issues with Oversize pay load when creating job*
>
> Once I was able to create contexts, creating jobs started throwing errors
> like:
>
> akka.remote.OversizedPayloadException: Discarding oversized payload [...] max allowed size 128000 bytes
>
> Adding the following line in the Job Server configuration got rid of this
> problem:
>
> akka.remote.netty.tcp.maximum-frame-size = 256000
>
> Now, my knowledge in Akka is limited so I'm not sure if that's a
> recommended thing to do or if that error might point towards something that
> is done wrong when creating the job.
>
> *Unable to run tests for assembly build*
>
> When building the JAR, I had to disable the tests in assembly because a
> couple of them have compilation errors (JobManagerActorSpec and
> JobStatusActorSpec). I understand that you didn't work out yet how to set
> up the tests for the multi-JVM side of things, but it might be good to fix
> the existing tests in the build - or to explicitly ignore them for the time
> being, so that we can still run the rest of the tests in assembly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spark-jobserver/spark-jobserver/pull/120#issuecomment-102584516>
> .
>



-- 
If you are free, you need to free somebody else.
If you have some power, then your job is to empower somebody else.
   --- Toni Morrison

Ankit1010 commented:
Thanks for the feedback Pierre! I have fixed the issues you brought up.

The akka messages do get bigger so i have added the frame-size settings.
I've moved the pidfile check from setenv to server_start.sh because individual context processes don't need pidfile checks.
I have updated all the tests to be inline with the changes and they pass on my end.

Ill try to write up tests for the new functionality and setup multi-jvm testing when I have some time. Is there anything else blocking this being merged?

velvia commented:
Ankit, thanks.  Not really, just need to have time to try it out myself.
The other option is to merge it to a branch, but not sure that's any better.

On Fri, May 29, 2015 at 5:25 PM, Ankit Soni <notifications@github.com>
wrote:

> Thanks for the feedback Pierre! I have fixed the issues you brought up.
>
> The akka messages do get bigger so i have added the frame-size settings.
> I've moved the pidfile check from setenv to server_start.sh because
> individual context processes don't need pidfile checks.
> I have updated all the tests to be inline with the changes and they pass
> on my end.
>
> Ill try to write up tests for the new functionality and setup multi-jvm
> testing when I have some time. Is there anything else blocking this being
> merged?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spark-jobserver/spark-jobserver/pull/120#issuecomment-106966546>
> .
>



-- 
If you are free, you need to free somebody else.
If you have some power, then your job is to empower somebody else.
   --- Toni Morrison

PierreVincent commented on job-server/src/main/resources/application.conf:
```diff
@@ -77,4 +99,8 @@ spray.can.server {
+}
+
+deploy {
+  manager-start-cmd = "./bin/manager_start.sh"
```
Once deployed, this file is in the root folder of the Job Server, not in /bin, so this should probably be:

```
manager-start-cmd = "./manager_start.sh"
```

This is still not optimal to have a relative path however, as it means that the server_start.sh script needs to be called from the folder where the Job Server is, otherwise the same error will occur. So for example it's not possible to start the Job Server with ```/opt/job-server/server_start.sh```.

It might be sufficient for now, as one should be able to override this property in the deployed configuration to set an absolute path to the manager_start.sh if necessary.

PierreVincent commented:
Thanks for all the updates @Ankit1010, I've rebuilt the JAR and tested it locally with Spark 1.3 and it seems to be working fine. The only thing I had to fix was the path to the manager_start.sh, see my comment above.

I see there is quite a large number of commits that have come through in master since you started your work on this, it might be worthwhile rebasing so we can verify that everything that was changed since is not clashing with your changes?

PierreVincent commented:
Hi @Ankit1010 and @velvia,

Just checking back if this PR has been parked for now or is it still being considered?

Thanks!

velvia commented:
Definitely still being considered, in fact one of the top priorities.
Just need time.  :-p

On Thu, Jun 25, 2015 at 4:42 AM, Pierre Vincent <notifications@github.com>
wrote:

> Hi @Ankit1010 <https://github.com/Ankit1010> and @velvia
> <https://github.com/velvia>,
>
> Just checking back if this PR has been parked for now or is it still being
> considered?
>
> Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spark-jobserver/spark-jobserver/pull/120#issuecomment-115217746>
> .
>



-- 
If you are free, you need to free somebody else.
If you have some power, then your job is to empower somebody else.
   --- Toni Morrison

Ankit1010 commented:
@PierreVincent i am also away for about 2 weeks. i suppose next steps are to fix the broken tests. add multi-jvm tests and rebase. Looks like i also pulled in some other unrelated jar caching changes in the last rebase that i should fix (sorry my git-fu is weak).

parvoberoi commented:
@PierreVincent: you mentioned that you were able to patch the diff and get it to work locally. I tried patching it based on the revision that this diff was based but the patch does not cleanly apply and almost all files lead to a merge conflict. Could you list down the steps you undertook to get it to work locally.

Thanks,
Parv.

PierreVincent commented:
@parvoberoi: I didn't patch the diff, I was running the Job Server from the source branch of the Pull Request (AppLovin:multi-jvm-2)

JoeYang commented:
Any updates on this? I am having trouble running 2 jobs under different context.

lzueclipse commented:
@PierreVincent: I checked out this branch and deployed it on a 3 nodes yarn cluster. While, it doesn't work. Followings are steps how I deployed the spark job server:
1) download the codes
2) create a local.sh based on the local.template.sh under spark-jobserver/config directory
3) execute spark-jobserver/bin/server_deploy.sh local.sh to deploy the spark job server to different nodes I configured in local.sh
4) execute sparkjobserver/server_start.sh to start the job server 
    to make server_start.sh script work,
     a) I manually copied settings.sh from bin to its parents folder
     b) I modified some paths in setenv.sh 
     c) I created a local.conf file to tell job server use manager-start.sh script as the value of parameter manager-start-cmd  and use yarn mode

5) create a context to test the job server, while the context created job can't finish. I checked the manager_start.log created by job server and found: 
     Error: Exception thrown by the agent : java.rmi.server.ExportException: Port already in use: 9999; nested exception is:
        java.net.BindException: Address already in use

Can you please tell me whether there are some extra steps when deploying the job server and what  caused the above exceptions?

velvia commented:
Hey guys, the plan for this PR is to move the code to an official branch and update it with the latest code.
Also this is related to the HA discussion which is now at #42 .
Timing: probably start this effort in earnest towards the end of this month.

noorul commented:
@Ankit1010 Are you by any chance using this patch at your side in production?

koettert commented:
Hello everyone,
we just started to adopt Spark job server for our project and contributed the ssl and Shiro based authentication. We are also very interested supporting multiple contexts and the work you did so far is great. 
Now that the job server supports authentication it would be awesome if a context could be assigned to a single user if authentication is enabled. The context would be assigned to the user that created it and only this user can work with the context e.g. start jobs etc. The context runs with the rights of the assigned user e.g. he could only access files in HDFS or read from Hive tables he has access to. What do you think about this suggestion? How can we help to advance this feature?
Thanks
Tobias

velvia commented:
Tobias,

Thanks very much.  After next week, I plan to move the PR into an official branch of the repo, and try to update it with the latest code.
Then it will be easier to work on together.

Some design questions:
- How should the context JVM be launched?   I’m thinking with Spark submit.

-Evan

> On Sep 16, 2015, at 11:45 PM, Tobias Koetter <notifications@github.com> wrote:
> 
> Hello everyone,
> we just started to adopt Spark job server for our project and contributed the ssl and Shiro based authentication. We are also very interested supporting multiple contexts and the work you did so far is great. 
> Now that the job server supports authentication it would be awesome if a context could be assigned to a single user if authentication is enabled. The context would be assigned to the user that created it and only this user can work with the context e.g. start jobs etc. The context runs with the rights of the assigned user e.g. he could only access files in HDFS or read from Hive tables he has access to. What do you think about this suggestion? How can we help to advance this feature?
> Thanks
> Tobias
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/spark-jobserver/spark-jobserver/pull/120#issuecomment-140984238>.
>

koettert commented:
Hi Evan,

perfect. We will start to help you as soon as the branch is created.
> - How should the context JVM be launched? I’m thinking with Spark submit.
I guess so to but haven't thought about it yet.

Bye,
Tobias


Am 17.09.2015 um 09:44 schrieb Evan Chan:
> Tobias,
> 
> Thanks very much. After next week, I plan to move the PR into an
> official branch of the repo, and try to update it with the latest code.
> Then it will be easier to work on together.
> 
> Some design questions:
> - How should the context JVM be launched? I’m thinking with Spark submit.
> 
> -Evan
> 
>> On Sep 16, 2015, at 11:45 PM, Tobias Koetter
> <notifications@github.com> wrote:
>>
>> Hello everyone,
>> we just started to adopt Spark job server for our project and
> contributed the ssl and Shiro based authentication. We are also very
> interested supporting multiple contexts and the work you did so far is
> great.
>> Now that the job server supports authentication it would be awesome if
> a context could be assigned to a single user if authentication is
> enabled. The context would be assigned to the user that created it and
> only this user can work with the context e.g. start jobs etc. The
> context runs with the rights of the assigned user e.g. he could only
> access files in HDFS or read from Hive tables he has access to. What do
> you think about this suggestion? How can we help to advance this feature?
>> Thanks
>> Tobias
>>
>> —
>> Reply to this email directly or view it on GitHub
> <https://github.com/spark-jobserver/spark-jobserver/pull/120#issuecomment-140984238>.
>>
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spark-jobserver/spark-jobserver/pull/120#issuecomment-140994201>.
>

noorul commented:
Tobias Koetter <notifications@github.com> writes:

> Hi Evan,
>
> perfect. We will start to help you as soon as the branch is created.
>> - How should the context JVM be launched? I’m thinking with Spark submit.
> I guess so to but haven't thought about it yet.
>

I am super excited about this work. Let me know if I can be of any help.

Thanks and Regards
Noorul

>
>
> Am 17.09.2015 um 09:44 schrieb Evan Chan:
>> Tobias,
>> 
>> Thanks very much. After next week, I plan to move the PR into an
>> official branch of the repo, and try to update it with the latest code.
>> Then it will be easier to work on together.
>> 
>> Some design questions:
>> - How should the context JVM be launched? I’m thinking with Spark submit.
>> 
>> -Evan
>> 
>>> On Sep 16, 2015, at 11:45 PM, Tobias Koetter
>> <notifications@github.com> wrote:
>>>
>>> Hello everyone,
>>> we just started to adopt Spark job server for our project and
>> contributed the ssl and Shiro based authentication. We are also very
>> interested supporting multiple contexts and the work you did so far is
>> great.
>>> Now that the job server supports authentication it would be awesome if
>> a context could be assigned to a single user if authentication is
>> enabled. The context would be assigned to the user that created it and
>> only this user can work with the context e.g. start jobs etc. The
>> context runs with the rights of the assigned user e.g. he could only
>> access files in HDFS or read from Hive tables he has access to. What do
>> you think about this suggestion? How can we help to advance this feature?
>>> Thanks
>>> Tobias
>>>
>>> —
>>> Reply to this email directly or view it on GitHub
>> <https://github.com/spark-jobserver/spark-jobserver/pull/120#issuecomment-140984238>.
>>>
>> 
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/spark-jobserver/spark-jobserver/pull/120#issuecomment-140994201>.
>> 
>
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/spark-jobserver/spark-jobserver/pull/120#issuecomment-140994993

Ankit1010 commented:
@noorul Yes, its working for us. Honestly this PR got very messy, but if you tweak the deployment scripts in bin/ a little bit it should work.

@velvia Sorry i don't have more time for this, but I did get this working with spark-submit. Here is the script to start a context for example:

    #!/bin/bash
    # Script to start the job manager
    # args: <job manager actor name> <cluster address>
    set -e
    
    get_abs_script_path() {
      ...
    }
    
    get_abs_script_path
    
    GC_OPTS="... "
    
    JAVA_OPTS="..."
    
    MAIN="spark.jobserver.JobManager"
    
    . $appdir/setenv.sh
    
    $SPARK_HOME/bin/spark-submit --class $MAIN --driver-memory $DRIVER_MEMORY \
        --conf "spark.executor.extraJavaOptions=$LOGGING_OPTS" \
        --driver--java-options "$GC_OPTS $JAVA_OPTS $LOGGING_OPTS $CONFIG_OVERRIDES" \
        ${@:3} $appdir/spark-job-server.jar $1 $2 $conffile &>> $LOG_DIR/manager_start.log &

andyramstad commented:
I'm pretty new to Spark and Scala, but I'm Java Developer.  

I'm running spark 1.5.0 on my VM and I have the job server installed on the same VM as my spark instance. 

I want to be able to submit multiple jobs in different SparkContexts, so I have cloned the from https://github.com/AppLovin/spark-jobserver and I had to tinker with the scripts to get it to work.

If I run against version v0.52 everything works fine.  

I would say I can get probably 20% of the jobs I submit via curl to actually execute, every other time I get one of the two stack traces below.  The same behavior that was exhibited in this ticket https://github.com/spark-jobserver/spark-jobserver/issues/206

I know this is a working branch, I just wanted to provide some feedback and would be willing to help test this as I don't want to re-invent the wheel as SJS provides a lot of the capability I'm after.  I'm looking forward to when this is in the core app.

1.
{
  "status": "ERROR",
  "result": {
    "message": "Ask timed out on [Actor[akka://JobServer/user/context-supervisor#-725810533]] after [15000 ms]",
    "errorClass": "akka.pattern.AskTimeoutException",
    "stack": ["akka.pattern.PromiseActorRef$$anonfun$1.apply$mcV$sp(AskSupport.scala:334)", "akka.actor.Scheduler$$anon$7.run(Scheduler.scala:117)", "scala.concurrent.Future$InternalCallbackExecutor$.scala$concurrent$Future$InternalCallbackExecutor$$unbatchedExecute(Future.scala:694)", "scala.concurrent.Future$InternalCallbackExecutor$.execute(Future.scala:691)", "akka.actor.LightArrayRevolverScheduler$TaskHolder.executeTask(Scheduler.scala:467)", "akka.actor.LightArrayRevolverScheduler$$anon$8.executeBucket$1(Scheduler.scala:419)", "akka.actor.LightArrayRevolverScheduler$$anon$8.nextTick(Scheduler.scala:423)", "akka.actor.LightArrayRevolverScheduler$$anon$8.run(Scheduler.scala:375)", "java.lang.Thread.run(Thread.java:745)"]
  }
}

2.
{
  "status": "ERROR",
  "result": {
    "message": "Futures timed out after [15 seconds]",
    "errorClass": "java.util.concurrent.TimeoutException",
    "stack": ["scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:219)", "scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:223)", "scala.concurrent.Await$$anonfun$result$1.apply(package.scala:107)", "akka.dispatch.MonitorableThreadFactory$AkkaForkJoinWorkerThread$$anon$3.block(ThreadPoolBuilder.scala:169)", "scala.concurrent.forkjoin.ForkJoinPool.managedBlock(ForkJoinPool.java:3640)", "akka.dispatch.MonitorableThreadFactory$AkkaForkJoinWorkerThread.blockOn(ThreadPoolBuilder.scala:167)", "scala.concurrent.Await$.result(package.scala:107)", "spark.jobserver.WebApi.spark$jobserver$WebApi$$getJobManagerForContext(WebApi.scala:401)", "spark.jobserver.WebApi$$anonfun$jobRoutes$5$$anonfun$apply$24.apply(WebApi.scala:299)", "spark.jobserver.WebApi$$anonfun$jobRoutes$5$$anonfun$apply$24.apply(WebApi.scala:292)", "spray.routing.ApplyConverterInstances$$anon$18$$anonfun$apply$5.apply(ApplyConverterInstances.scala:53)", "spray.routing.ApplyConverterInstances$$anon$18$$anonfun$apply$5.apply(ApplyConverterInstances.scala:52)", "spray.routing.ConjunctionMagnet$$anon$1$$anon$2$$anonfun$happly$1$$anonfun$apply$1.apply(Directive.scala:38)", "spray.routing.ConjunctionMagnet$$anon$1$$anon$2$$anonfun$happly$1$$anonfun$apply$1.apply(Directive.scala:37)", "spray.routing.directives.BasicDirectives$$anon$2.happly(BasicDirectives.scala:79)", "spray.routing.Directive$$anon$7$$anonfun$happly$4.apply(Directive.scala:86)", "spray.routing.Directive$$anon$7$$anonfun$happly$4.apply(Directive.scala:86)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anonfun$mapRequestContext$1$$anonfun$apply$1.apply(BasicDirectives.scala:30)", "spray.routing.directives.BasicDirectives$$anonfun$mapRequestContext$1$$anonfun$apply$1.apply(BasicDirectives.scala:30)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anonfun$mapRequestContext$1$$anonfun$apply$1.apply(BasicDirectives.scala:30)", "spray.routing.directives.BasicDirectives$$anonfun$mapRequestContext$1$$anonfun$apply$1.apply(BasicDirectives.scala:30)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1$$anonfun$apply$1.apply(RouteConcatenation.scala:32)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1$$anonfun$apply$1.apply(RouteConcatenation.scala:31)", "spray.routing.RequestContext$$anonfun$withRejectionHandling$1.applyOrElse(RequestContext.scala:130)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply$mcVL$sp(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:25)", "spray.routing.RequestContext$$anon$1.handle(RequestContext.scala:84)", "akka.spray.UnregisteredActorRefBase.$bang(UnregisteredActorRefBase.scala:72)", "spray.routing.RequestContext$$anon$3.handle(RequestContext.scala:102)", "akka.spray.UnregisteredActorRefBase.$bang(UnregisteredActorRefBase.scala:72)", "spray.routing.RequestContext.reject(RequestContext.scala:202)", "spray.routing.directives.RouteDirectives$$anon$1.apply(RouteDirectives.scala:35)", "spray.routing.directives.RouteDirectives$$anon$1.apply(RouteDirectives.scala:34)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1$$anonfun$apply$1.apply(RouteConcatenation.scala:32)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1$$anonfun$apply$1.apply(RouteConcatenation.scala:31)", "spray.routing.RequestContext$$anonfun$withRejectionHandling$1.applyOrElse(RequestContext.scala:130)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply$mcVL$sp(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:25)", "spray.routing.RequestContext$$anon$1.handle(RequestContext.scala:84)", "akka.spray.UnregisteredActorRefBase.$bang(UnregisteredActorRefBase.scala:72)", "spray.routing.RequestContext$$anon$3.handle(RequestContext.scala:102)", "akka.spray.UnregisteredActorRefBase.$bang(UnregisteredActorRefBase.scala:72)", "spray.routing.RequestContext.reject(RequestContext.scala:202)", "spray.routing.directives.RouteDirectives$$anon$1.apply(RouteDirectives.scala:35)", "spray.routing.directives.RouteDirectives$$anon$1.apply(RouteDirectives.scala:34)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1$$anonfun$apply$1.apply(RouteConcatenation.scala:32)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1$$anonfun$apply$1.apply(RouteConcatenation.scala:31)", "spray.routing.RequestContext$$anonfun$withRejectionHandling$1.applyOrElse(RequestContext.scala:130)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply$mcVL$sp(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:25)", "spray.routing.RequestContext$$anon$1.handle(RequestContext.scala:84)", "akka.spray.UnregisteredActorRefBase.$bang(UnregisteredActorRefBase.scala:72)", "spray.routing.RequestContext$$anon$3.handle(RequestContext.scala:102)", "akka.spray.UnregisteredActorRefBase.$bang(UnregisteredActorRefBase.scala:72)", "spray.routing.RequestContext.reject(RequestContext.scala:202)", "spray.routing.directives.RouteDirectives$$anon$1.apply(RouteDirectives.scala:35)", "spray.routing.directives.RouteDirectives$$anon$1.apply(RouteDirectives.scala:34)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1$$anonfun$apply$1.apply(RouteConcatenation.scala:32)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1$$anonfun$apply$1.apply(RouteConcatenation.scala:31)", "spray.routing.RequestContext$$anonfun$withRejectionHandling$1.applyOrElse(RequestContext.scala:130)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply$mcVL$sp(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:25)", "spray.routing.RequestContext$$anon$1.handle(RequestContext.scala:84)", "akka.spray.UnregisteredActorRefBase.$bang(UnregisteredActorRefBase.scala:72)", "spray.routing.RequestContext.reject(RequestContext.scala:202)", "spray.routing.directives.RouteDirectives$$anon$1.apply(RouteDirectives.scala:35)", "spray.routing.directives.RouteDirectives$$anon$1.apply(RouteDirectives.scala:34)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1.apply(RouteConcatenation.scala:30)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1.apply(RouteConcatenation.scala:29)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1.apply(RouteConcatenation.scala:30)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1.apply(RouteConcatenation.scala:29)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1.apply(RouteConcatenation.scala:30)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1.apply(RouteConcatenation.scala:29)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1.apply(RouteConcatenation.scala:30)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1.apply(RouteConcatenation.scala:29)", "spray.routing.directives.BasicDirectives$$anonfun$mapRequestContext$1$$anonfun$apply$1.apply(BasicDirectives.scala:30)", "spray.routing.directives.BasicDirectives$$anonfun$mapRequestContext$1$$anonfun$apply$1.apply(BasicDirectives.scala:30)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1$$anonfun$apply$1.apply(RouteConcatenation.scala:32)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1$$anonfun$apply$1.apply(RouteConcatenation.scala:31)", "spray.routing.RequestContext$$anonfun$withRejectionHandling$1.applyOrElse(RequestContext.scala:130)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply$mcVL$sp(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:25)", "spray.routing.RequestContext$$anon$1.handle(RequestContext.scala:84)", "akka.spray.UnregisteredActorRefBase.$bang(UnregisteredActorRefBase.scala:72)", "spray.routing.RequestContext$$anon$3.handle(RequestContext.scala:102)", "akka.spray.UnregisteredActorRefBase.$bang(UnregisteredActorRefBase.scala:72)", "spray.routing.RequestContext.reject(RequestContext.scala:202)", "spray.routing.directives.RouteDirectives$$anon$5.apply(RouteDirectives.scala:63)", "spray.routing.directives.RouteDirectives$$anon$5.apply(RouteDirectives.scala:62)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1$$anonfun$apply$1.apply(RouteConcatenation.scala:32)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1$$anonfun$apply$1.apply(RouteConcatenation.scala:31)", "spray.routing.RequestContext$$anonfun$withRejectionHandling$1.applyOrElse(RequestContext.scala:130)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply$mcVL$sp(AbstractPartialFunction.scala:33)", "scala.runtime.Abstrac100 15357  100 15357    0     0   1022      0  0:00:15  0:00:15 --:--:--  4034xt.scala:84)", "akka.spray.UnregisteredActorRefBase.$bang(UnregisteredActorRefBase.scala:72)", "spray.routing.RequestContext.reject(RequestContext.scala:202)", "spray.routing.directives.RouteDirectives$$anon$5.apply(RouteDirectives.scala:63)", "spray.routing.directives.RouteDirectives$$anon$5.apply(RouteDirectives.scala:62)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.directives.BasicDirectives$$anon$3$$anonfun$happly$1.apply(BasicDirectives.scala:92)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1.apply(RouteConcatenation.scala:30)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1.apply(RouteConcatenation.scala:29)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1.apply(RouteConcatenation.scala:30)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1.apply(RouteConcatenation.scala:29)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1.apply(RouteConcatenation.scala:30)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1.apply(RouteConcatenation.scala:29)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1.apply(RouteConcatenation.scala:30)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1.apply(RouteConcatenation.scala:29)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1.apply(RouteConcatenation.scala:30)", "spray.routing.RouteConcatenation$RouteConcatenation$$anonfun$$tilde$1.apply(RouteConcatenation.scala:29)", "spray.routing.directives.BasicDirectives$$anonfun$mapRequestContext$1$$anonfun$apply$1.apply(BasicDirectives.scala:30)", "spray.routing.directives.BasicDirectives$$anonfun$mapRequestContext$1$$anonfun$apply$1.apply(BasicDirectives.scala:30)", "spray.routing.directives.ExecutionDirectives$$anonfun$handleExceptions$1$$anonfun$apply$4.apply(ExecutionDirectives.scala:35)", "spray.routing.directives.ExecutionDirectives$$anonfun$handleExceptions$1$$anonfun$apply$4.apply(ExecutionDirectives.scala:33)", "spray.routing.HttpServiceBase$class.runSealedRoute$1(HttpService.scala:36)", "spray.routing.HttpServiceBase$$anonfun$runRoute$1.applyOrElse(HttpService.scala:46)", "akka.actor.Actor$class.aroundReceive(Actor.scala:467)", "spray.routing.SimpleRoutingApp$$anonfun$1$$anon$1.aroundReceive(SimpleRoutingApp.scala:54)", "akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)", "akka.actor.ActorCell.invoke(ActorCell.scala:487)", "akka.dispatch.Mailbox.processMailbox(Mailbox.scala:238)", "akka.dispatch.Mailbox.run(Mailbox.scala:220)", "akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:397)", "scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)", "scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)", "scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)", "scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)"]
  }
}

wlsailor commented:
when will job server support one jvm per context, is this in schedule? We really need the feature, but this pull request has too many conflict to merge by ourselves, eager to see this feature in next version, thanks to your work.

velvia commented:
Leighton, I’m working on merging and getting this working.   it will appear in an official branch with the latest code soon, for everyone to test.  At that time I’ll also close out this PR.

-Evan

> On Nov 5, 2015, at 12:33 AM, Leighton Wong <notifications@github.com> wrote:
> 
> when will job server support one jvm per context, is this in schedule? We really need the feature, but this pull request has too many conflict to merge by ourselves, eager to see this feature in next version, thanks to your work.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/spark-jobserver/spark-jobserver/pull/120#issuecomment-153987535>.
>

velvia commented:
@wlsailor @Ankit1010 @PierreVincent @noorul @koettert and others:  I have merged this PR into an official branch, `feature/context-per-jvm`, with close to the latest code on master, Spark 1.5.1 and 0.6.1-SNAPSHOT.  Most of the tests pass, but I have not had time to deploy and try it.    Thus the changes between this branch and what's in the new branch:

* All the changes in master, including Spark 1.5.1, streaming, SQL contexts, HTTPS/auth/Shiro, and too many more to be listed
* The `manager_start.sh` script uses spark-submit
* `JobInfoActor` was reverted to use plain `JobDAO` and not the actor, since it is all local, and simplifies the code and merging (and also eliminates need to fiddle with Akka frame size)

Please try it out you guys!   This PR will be closed soon since the new code is available.

wlsailor commented:
Thanks very much, we will test deploying right now.

wlsailor commented:
@velvia Deploy successful after comment the last line of setenv.sh, no compute-classpath.sh any more for recent spark release, but we can not create persistent context by a post request, return error:
```
{
  "status": "CONTEXT INIT ERROR",
  "result": {
    "message": "Failed to launch context process, got exit code 1",
    "errorClass": "java.io.IOException",
    "stack": [
      "spark.jobserver.AkkaClusterSupervisorActor$$anonfun$1.apply$mcV$sp(AkkaClusterSupervisorActor.scala:185)",
      "spark.jobserver.AkkaClusterSupervisorActor$$anonfun$1.apply(AkkaClusterSupervisorActor.scala:181)",
      "spark.jobserver.AkkaClusterSupervisorActor$$anonfun$1.apply(AkkaClusterSupervisorActor.scala:181)",
      "scala.util.Try$.apply(Try.scala:192)",
      "spark.jobserver.AkkaClusterSupervisorActor.spark$jobserver$AkkaClusterSupervisorActor$$startContext(AkkaClusterSupervisorActor.scala:181)",
      "spark.jobserver.AkkaClusterSupervisorActor$$anonfun$wrappedReceive$1.applyOrElse(AkkaClusterSupervisorActor.scala:99)",
      "scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)",
      "ooyala.common.akka.ActorStack$$anonfun$receive$1.applyOrElse(ActorStack.scala:33)",
      "scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)",
      "ooyala.common.akka.Slf4jLogging$$anonfun$receive$1$$anonfun$applyOrElse$1.apply$mcV$sp(Slf4jLogging.scala:26)",
      "ooyala.common.akka.Slf4jLogging$class.ooyala$common$akka$Slf4jLogging$$withAkkaSourceLogging(Slf4jLogging.scala:35)",
      "ooyala.common.akka.Slf4jLogging$$anonfun$receive$1.applyOrElse(Slf4jLogging.scala:25)",
      "scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)",
      "ooyala.common.akka.ActorMetrics$$anonfun$receive$1.applyOrElse(ActorMetrics.scala:24)",
      "akka.actor.Actor$class.aroundReceive(Actor.scala:467)",
      "ooyala.common.akka.InstrumentedActor.aroundReceive(InstrumentedActor.scala:8)",
      "akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)",
      "akka.actor.ActorCell.invoke(ActorCell.scala:487)",
      "akka.dispatch.Mailbox.processMailbox(Mailbox.scala:238)",
      "akka.dispatch.Mailbox.run(Mailbox.scala:220)",
      "akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:397)",
      "scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)",
      "scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)",
      "scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)",
      "scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)"
    ]
  }
}
```

PierreVincent commented:
@velvia, thanks, I will try it out as soon as possible. Will get back before the end of the week!

PierreVincent commented:
I am getting a similar error when starting a Job with ad-hoc context:

```
{
  "status": "ERROR",
  "result": {
    "errorClass": "java.lang.RuntimeException",
    "cause": "Failed to launch context process, got exit code 1",
    "stack": [
      "spark.jobserver.AkkaClusterSupervisorActor$$anonfun$1.apply$mcV$sp(AkkaClusterSupervisorActor.scala:185)",
      "spark.jobserver.AkkaClusterSupervisorActor$$anonfun$1.apply(AkkaClusterSupervisorActor.scala:181)",
      "spark.jobserver.AkkaClusterSupervisorActor$$anonfun$1.apply(AkkaClusterSupervisorActor.scala:181)",
      "scala.util.Try$.apply(Try.scala:161)",
      "spark.jobserver.AkkaClusterSupervisorActor.spark$jobserver$AkkaClusterSupervisorActor$$startContext(AkkaClusterSupervisorActor.scala:181)",
      "spark.jobserver.AkkaClusterSupervisorActor$$anonfun$wrappedReceive$1.applyOrElse(AkkaClusterSupervisorActor.scala:115)",
      "scala.runtime.AbstractPartialFunction$mcVL$sp.apply$mcVL$sp(AbstractPartialFunction.scala:33)",
      "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:33)",
      "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:25)",
      "ooyala.common.akka.ActorStack$$anonfun$receive$1.applyOrElse(ActorStack.scala:33)",
      "scala.runtime.AbstractPartialFunction$mcVL$sp.apply$mcVL$sp(AbstractPartialFunction.scala:33)",
      "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:33)",
      "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:25)",
      "ooyala.common.akka.Slf4jLogging$$anonfun$receive$1$$anonfun$applyOrElse$1.apply$mcV$sp(Slf4jLogging.scala:26)",
      "ooyala.common.akka.Slf4jLogging$class.ooyala$common$akka$Slf4jLogging$$withAkkaSourceLogging(Slf4jLogging.scala:35)",
      "ooyala.common.akka.Slf4jLogging$$anonfun$receive$1.applyOrElse(Slf4jLogging.scala:25)",
      "scala.runtime.AbstractPartialFunction$mcVL$sp.apply$mcVL$sp(AbstractPartialFunction.scala:33)",
      "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:33)",
      "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:25)",
      "ooyala.common.akka.ActorMetrics$$anonfun$receive$1.applyOrElse(ActorMetrics.scala:24)",
      "akka.actor.Actor$class.aroundReceive(Actor.scala:465)",
      "ooyala.common.akka.InstrumentedActor.aroundReceive(InstrumentedActor.scala:8)",
      "akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)",
      "akka.actor.ActorCell.invoke(ActorCell.scala:487)",
      "akka.dispatch.Mailbox.processMailbox(Mailbox.scala:238)",
      "akka.dispatch.Mailbox.run(Mailbox.scala:220)",
      "akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:393)",
      "scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)",
      "scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)",
      "scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)",
      "scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)"
    ],
    "causingClass": "java.io.IOException",
    "message": "java.io.IOException: Failed to launch context process, got exit code 1"
  }
}
```

velvia commented:
@PierreVincent @wlsailor thanks for trying. You are right, I forgot to leave out `compute_classpath.sh`.  Will merge #309, thanks for the PR.

wlsailor commented:
@velvia Sorry ,I forgot to commit the change for removing compute_classpath.sh in setenv.sh #309 .

velvia commented:
NP, I just removed it and pushed the change.

> On Nov 10, 2015, at 8:08 PM, Leighton Wong <notifications@github.com> wrote:
> 
> @velvia <https://github.com/velvia> Sorry ,I forgot to commit the change for removing compute_classpath.sh in setenv.sh #309 <https://github.com/spark-jobserver/spark-jobserver/pull/309> .
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/spark-jobserver/spark-jobserver/pull/120#issuecomment-155656107>.
>

PierreVincent commented:
I've tried the latest changes and this is looking good.

I've been able to start multiple jobs simultaneously on separate ad-hoc contexts, Spark is no longer complaining about sharing JVM and more importantly the jobs succeed!

Thanks for the great work guys!

velvia commented:
@PierreVincent awesome!

I'm going to close this PR now.   We can open another one for the new branch.   TODOs for the new branch:
- Make all tests pass
- A switch so folks can fall back to the local everything-in-one-JVM mode (and maybe make that the default for local SBT development?)
- Documentation

PierreVincent commented:
Cool, don't hesitate to tag me when you open a new PR so I can validate it on my end.

salihatilay commented:
Hello, I want to create 2 streaming context,
For context-per-jvm, as stated in the documentation, i switched job-server-extras,
After i restart it, In order to create a new Context, posting below
http://localhost:8090/contexts/xx?context-factory=spark.jobserver.context.StreamingContextFactory'.
I am getting  "./manager-start.sh" is not found. Then i changed the path to "../bin/manager-start.sh.", but this time i am getting following error. and logs ..../settings.sh not found. What i am doing wrong ?

{
  "status": "CONTEXT INIT ERROR",
  "result": {
    "message": "Failed to launch context process, got exit code 1",
    "errorClass": "java.io.IOException",
    "stack": ["spark.jobserver.AkkaClusterSupervisorActor$$anonfun$1.apply$mcV$sp(AkkaClusterSupervisorActor.scala:191)", "spark.jobserver.AkkaClusterSupervisorActor$$anonfun$1.apply(AkkaClusterSupervisorActor.scala:187)", "spark.jobserver.AkkaClusterSupervisorActor$$anonfun$1.apply(AkkaClusterSupervisorActor.scala:187)", "scala.util.Try$.apply(Try.scala:161)", "spark.jobserver.AkkaClusterSupervisorActor.spark$jobserver$AkkaClusterSupervisorActor$$startContext(AkkaClusterSupervisorActor.scala:187)", "spark.jobserver.AkkaClusterSupervisorActor$$anonfun$wrappedReceive$1.applyOrElse(AkkaClusterSupervisorActor.scala:100)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply$mcVL$sp(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:25)", "ooyala.common.akka.ActorStack$$anonfun$receive$1.applyOrElse(ActorStack.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply$mcVL$sp(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:25)", "ooyala.common.akka.Slf4jLogging$$anonfun$receive$1$$anonfun$applyOrElse$1.apply$mcV$sp(Slf4jLogging.scala:26)", "ooyala.common.akka.Slf4jLogging$class.ooyala$common$akka$Slf4jLogging$$withAkkaSourceLogging(Slf4jLogging.scala:35)", "ooyala.common.akka.Slf4jLogging$$anonfun$receive$1.applyOrElse(Slf4jLogging.scala:25)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply$mcVL$sp(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:25)", "ooyala.common.akka.ActorMetrics$$anonfun$receive$1.applyOrElse(ActorMetrics.scala:24)", "akka.actor.Actor$class.aroundReceive(Actor.scala:467)", "ooyala.common.akka.InstrumentedActor.aroundReceive(InstrumentedActor.scala:8)", "akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)", "akka.actor.ActorCell.invoke(ActorCell.scala:487)", "akka.dispatch.Mailbox.processMailbox(Mailbox.scala:238)", "akka.dispatch.Mailbox.run(Mailbox.scala:220)", "akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:397)", "scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)", "scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)", "scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)", "scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)"]
  }
}

velvia commented:
Are you trying to start it from sbt?  I think right now It assumes you have deployed. 

-Evan
"Never doubt that a small group of thoughtful, committed citizens can change the world" - M. Mead

> On Nov 29, 2015, at 7:09 AM, salihatilay <notifications@github.com> wrote:
> 
> Hello, I want to create 2 streaming context,
> For context-per-jvm, as stated in the documentation, i switched job-server-extras,
> After i restart it, In order to create a new Context, posting below
> http://localhost:8090/contexts/xx?context-factory=spark.jobserver.context.StreamingContextFactory'.
> I am getting "./manager-start.sh" is not found. Then i changed the path to "../bin/manager-start.sh.", but this time i am getting following error. and logs ..../settings.sh not found. What i am doing wrong ?
> 
> {
> "status": "CONTEXT INIT ERROR",
> "result": {
> "message": "Failed to launch context process, got exit code 1",
> "errorClass": "java.io.IOException",
> "stack": ["spark.jobserver.AkkaClusterSupervisorActor$$anonfun$1.apply$mcV$sp(AkkaClusterSupervisorActor.scala:191)", "spark.jobserver.AkkaClusterSupervisorActor$$anonfun$1.apply(AkkaClusterSupervisorActor.scala:187)", "spark.jobserver.AkkaClusterSupervisorActor$$anonfun$1.apply(AkkaClusterSupervisorActor.scala:187)", "scala.util.Try$.apply(Try.scala:161)", "spark.jobserver.AkkaClusterSupervisorActor.spark$jobserver$AkkaClusterSupervisorActor$$startContext(AkkaClusterSupervisorActor.scala:187)", "spark.jobserver.AkkaClusterSupervisorActor$$anonfun$wrappedReceive$1.applyOrElse(AkkaClusterSupervisorActor.scala:100)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply$mcVL$sp(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:25)", "ooyala.common.akka.ActorStack$$anonfun$receive$1.applyOrElse(ActorStack.scala:33)", "scala.runt ime.AbstractPartialFunction$mcVL$sp.apply$mcVL$sp(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:25)", "ooyala.common.akka.Slf4jLogging$$anonfun$receive$1$$anonfun$applyOrElse$1.apply$mcV$sp(Slf4jLogging.scala:26)", "ooyala.common.akka.Slf4jLogging$class.ooyala$common$akka$Slf4jLogging$$withAkkaSourceLogging(Slf4jLogging.scala:35)", "ooyala.common.akka.Slf4jLogging$$anonfun$receive$1.applyOrElse(Slf4jLogging.scala:25)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply$mcVL$sp(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:33)", "scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:25)", "ooyala.common.akka.ActorMetrics$$anonfun$receive$1.applyOrElse(ActorMetrics.scala:24)", "akka.actor.Actor$class.aroundReceive(Actor.scal a:467)", "ooyala.common.akka.InstrumentedActor.aroundReceive(InstrumentedActor.scala:8)", "akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)", "akka.actor.ActorCell.invoke(ActorCell.scala:487)", "akka.dispatch.Mailbox.processMailbox(Mailbox.scala:238)", "akka.dispatch.Mailbox.run(Mailbox.scala:220)", "akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:397)", "scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)", "scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)", "scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)", "scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)"]
> }
> }
> 
> —
> Reply to this email directly or view it on GitHub.
>

salihatilay commented:
Thanks for reply,
Yes, i started it from sbt. So for now the only way is to deploy it to Spark ?

velvia commented:
For now yeah.  I haven't tried the sbt method but I think shouldn't be too hard to fix. 

-Evan
"Never doubt that a small group of thoughtful, committed citizens can change the world" - M. Mead

> On Nov 29, 2015, at 8:44 AM, salihatilay <notifications@github.com> wrote:
> 
> Thanks for reply,
> Yes, i started it from sbt. So for now the only way is to deploy it to Spark ?
> 
> —
> Reply to this email directly or view it on GitHub.
>
